pkg/manager/volumes/delegation/azure/disk_modifier.go

```diff
@@ -0,0 +1,308 @@
+		return nil, fmt.Errorf("failed to get Azure disk info from PV: %v", err)
+	}
+
+	diskResponse, err := m.DiskClient.Get(ctx, resourceGroupName, diskName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get disk: %v", err)
+	}
+
+	disk := diskResponse.Disk
+	return &Volume{
+		VolumeId:   *disk.ID,
+		Size:       disk.Properties.DiskSizeGB,
+		IOPS:       disk.Properties.DiskIOPSReadWrite,
+		Throughput: disk.Properties.DiskMBpsReadWrite,
+		Type:       string(*disk.SKU.Name),
+	}, nil
+}
+
+func (m *AzureDiskModifier) getExpectedVolume(pvc *corev1.PersistentVolumeClaim, pv *corev1.PersistentVolume, sc *storagev1.StorageClass) (*Volume, error) {
+	v := Volume{}
+	if err := utilerrors.NewAggregate([]error{
+		m.setArgsFromPVC(&v, pvc),
+		m.setArgsFromPV(&v, pv),
+		m.setArgsFromStorageClass(&v, sc),
+	}); err != nil {
+		return nil, err
+	}
+
+	return &v, nil
+}
+
+func (m *AzureDiskModifier) MinWaitDuration() time.Duration {
+	return defaultWaitDuration
+}
+
+func (m *AzureDiskModifier) setArgsFromPVC(v *Volume, pvc *corev1.PersistentVolumeClaim) error {
+	size, err := getSizeFromPVC(pvc)
+	if err != nil {
+		return err
+	}
+	v.Size = ptr.To(int32(size))
+	return nil
+}
+
+func getSizeFromPVC(pvc *corev1.PersistentVolumeClaim) (int64, error) {
+	quantity := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+	sizeBytes := quantity.ScaledValue(0)
+	size := sizeBytes / 1024 / 1024 / 1024
+
+	if size < minSize || size > maxSize {
+		return 0, fmt.Errorf("invalid storage size: %v", quantity)
+	}
+	return size, nil
+}
+
+func (m *AzureDiskModifier) setArgsFromPV(v *Volume, pv *corev1.PersistentVolume) error {
+	v.VolumeId = pv.Spec.CSI.VolumeHandle
+	return nil
+}
+
+func (m *AzureDiskModifier) setArgsFromStorageClass(v *Volume, sc *storagev1.StorageClass) error {
+	if sc == nil {
+		return nil
+	}
+	throughput, err := getParamInt64(sc.Parameters, paramKeyThroughput)
+	if err != nil {
+		return err
+	}
+	v.Throughput = throughput
+
+	iops, err := getParamInt64(sc.Parameters, paramKeyIOPS)
+	if err != nil {
+		return err
+	}
+	v.IOPS = iops
+
+	v.Type = sc.Parameters[paramKeyType]
+	return nil
+}
+
+func getParamInt64(params map[string]string, key string) (*int64, error) {
+	str, ok := params[key]
+	if !ok || str == "" {
+		return nil, nil
+	}
+	param, err := strconv.ParseInt(str, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("can't parse %v param in storage class: %v", key, err)
+	}
+
+	return ptr.To(param), nil
+}
+
+func getDiskInfoFromVolumeID(volumeID string) (diskName string, subscriptionID string, resourceGroupName string, err error) {
+	// get diskName, subscriptionID, resourceGroupName from volumeHandle
+	// example: /subscriptions/xxxx/resourceGroups/xxxx/providers/Microsoft.Compute/disks/xxxx
+	parts := strings.Split(volumeID, "/")
+	if len(parts) < 9 {
+		return "", "", "", fmt.Errorf("invalid volumeHandle format")
+	}
+	subscriptionID = parts[2]
+	resourceGroupName = parts[4]
+	diskName = parts[len(parts)-1]
+
```

Comment:
Do we need to check extract `if len(parts) = 9` above, or should we use `strings.SplitN` hear?

Reply:
It should be set to != 9. I fixed it.